rbac/management/role/view.py

```diff
@@ -418,15 +440,45 @@ def update(self, request, *args, **kwargs):
         validate_uuid(kwargs.get("uuid"), "role uuid validation")
         self.validate_role(request)
 
-        role = self.get_object()
-        update_role = super().update(request=request, args=args, kwargs=kwargs)
+        update_role = self.update_with_relation_api_replication(request=request, args=args, kwargs=kwargs)
 
         if status.is_success(update_role.status_code):
             auditlog = AuditLog()
+            role = self.get_object()
             auditlog.log_edit(request, AuditLog.ROLE, role)
 
         return update_role
 
+    def update_with_relation_api_replication(self, request, *args, **kwargs):
+        """Update a role with replicating data into Relation API."""
+        try:
+            role = self.get_object()
+            dual_write_handler = RelationApiDualWriteHandler(role, "UPDATE")
+            dual_write_handler.generate_relations_from_current_state_of_role()
+            with transaction.atomic():
+                response = super().update(request=request, args=args, kwargs=kwargs)
+                dual_write_handler.generate_replication_event_to_outbox(self.get_object())
```

Comment:
So there is another race here that could get us dead in the water. Basically, all of the `RelationApiDualWriteHandler` operations need to be done within a single transaction that is either row level locked for the Role, or uses repeatable read isolation.

Annotated code with what the problem is:

```python
            # This is outside a tx, so uses current snapshot and does not block other transactions
            role = self.get_object()
            dual_write_handler = RelationApiDualWriteHandler(role, "UPDATE")
            # At this point, Role state may be changed in DB,
            # and another outbox event could have been written with different role state
            # now unknown to this method
            dual_write_handler.generate_relations_from_current_state_of_role()
            with transaction.atomic():
                # Now we lock the Role, but too late–the Role has already changed
                response = super().update(request=request, args=args, kwargs=kwargs)
                # Now we generate mappings based on update, but we do not remove the right relations
                # because those relations were generating concurrently.
                dual_write_handler.generate_replication_event_to_outbox(self.get_object())
```

Fixing this is the same thing as we talked about before – we need to bring the whole thing into the same transaction and either (a) blocking lock the row or (b) use repeatable read.

We discussed a bit about (a) in slack if we want to go down that route but I'm okay either way

Reply:
So this needs to be applied everywhere the dual write handler is used I think.